src/libopensc/log.c

```diff
@@ -84,10 +84,21 @@ void sc_do_log_openssl(sc_context_t *ctx, int level, const char *file, int line,
 	}
 	ERR_print_errors(bio);
 
-	if ((length = BIO_pending(bio)) <= 0 ||
-			(buffer = malloc(length)) == NULL ||
-			BIO_read(bio, buffer, length) <= 0) {
-		sc_do_log(ctx, level, file, line, func, "Cannot log OpenSSL error");
+	length = BIO_pending(bio);
+	if (length <= 0) {
+		/* no error? */
+		goto end;
+	}
+	/* trailing null byte */
+	buffer = malloc(length + 1);
+	if (buffer == NULL) {
+		sc_do_log(ctx, level, file, line, func, "No memory!");
+		goto end;
+	}
+	rc = BIO_read(bio, buffer, length);
+	buffer[length] = '\n';
+	if (rc <= 0) {
+		sc_do_log(ctx, level, file, line, func, "Cannot read OpenSSL error");
```

Comment:
I think it would be useful to keep the old error message "Cannot log OpenSSL error"

Reply:
The thing was that the same log message was here for different things and I find it useful when they are as specific as possible so one does not have to go through the code and line offsets to figure out which one of these same messages failed. The old message `Cannot log OpenSSL error` is still on line 82 (which I would consider to rewrite also to something about memory issues).

Reply:
thanks for the explanation 👍